Wasmtime: support a notion of "custom code publisher".

cfallin opened the pull request:
In some `no_std` environments, virtual memory usage is *generally* prohibited for performance-predictability reasons, but the MMU hardware is still in use for permissions (e.g., `W^X` write-xor-execute). Occasional changes to page mapping permissions are thus necessary when new modules are loaded dynamically, and are acceptable in that context. Wasmtime needs a way to support "publishing" code (making it executable) in such environments.

Rather than try to segment the `signals-based-traps` divide further, and piece out the code-publishing parts from the heap parts, and backdoor a path to `mprotect` in an otherwise `no_std` build, in this PR I have opted to add a trait an impl of which the embedder can provide to the `Config` to implement custom actions for "code publish". This otherwise operates properly in a
no-`signals-based-traps` environment, e.g., the module backing memory itself is regularly allocated rather than mmap'd (but is now aligned to the degree requested by the trait impl).

<!--
Please make sure you include the following information:

- If this work has been discussed elsewhere, please include a link to that
  conversation. If it was discussed in an issue, just mention "issue #...".

- Explain why this change is needed. If the details are in an issue already,
  this can be brief.

Our development process is documented in the Wasmtime book:
https://docs.wasmtime.dev/contributing-development-process.html

Please ensure all communication follows the code of conduct:
https://github.com/bytecodealliance/wasmtime/blob/main/CODE_OF_CONDUCT.md
-->

cfallin commented on crates/wasmtime/src/runtime/vm/mmap_vec.rs:
```diff
@@ -68,13 +94,49 @@ impl MmapVec {
+            assert!(alignment <= crate::runtime::vm::host_page_size());
+            return Ok(MmapVec::new_mmap(Mmap::with_at_least(size)?, size));
+        }
+        #[cfg(not(feature = "signals-based-traps"))]
+        {
+            return Ok(MmapVec::new_vec_aligned(
+                vec![0; size + alignment],
+                size,
+                alignment,
+            ));
+        }
```
Unfortunately that's not an option in my embedding: memory must come from the `alloc` allocator.

cfallin replied:
Ah, actually, do you mean the lower-level `alloc` APIs? I'll look into that.

cfallin replied:
Done!

cfallin commented on crates/wasmtime/Cargo.toml:
```diff
@@ -364,3 +364,14 @@ wave = ["dep:wasm-wave"]
+# mechanism to allow loading of modules and making their code
+# executable. This feature defines a trait that can be provided to the
+# Engine to permit this.
+custom-code-memory = []
```
Sure, happy to fold it in; was worried the extra checks might cause some consternation but you're right, it's probably not too bad (and loading modules is relatively not a hot path).

cfallin replied:
Done.

cfallin commented on crates/wasmtime/src/runtime/vm/mmap_vec.rs:
```diff
@@ -68,13 +94,49 @@ impl MmapVec {
+    /// A new `MmapVec` is allocated to hold the contents of `slice` and then
+    /// `slice` is copied into the new mmap. It's recommended to avoid this
+    /// method if possible to avoid the need to copy data around.  pub
+    pub fn from_slice_with_alignment(slice: &[u8], align: usize) -> Result<MmapVec> {
```
Done.

alexcrichton commented on crates/wasmtime/src/engine.rs:
```diff
@@ -61,6 +63,8 @@ struct EngineInner {
     signatures: TypeRegistry,
     #[cfg(feature = "runtime")]
     epoch: AtomicU64,
+    #[cfg(feature = "runtime")]
+    custom_code_memory: Option<Arc<dyn CustomCodeMemory>>,
```
This is already stored as-is in `Config` so might be able to access from there (stored above) instead of also storing here too?

cfallin replied:
Ah, right, good point -- updated.

alexcrichton commented on crates/wasmtime/src/engine.rs:
```diff
@@ -655,6 +661,11 @@ impl Engine {
     }
 
+    #[cfg(feature = "runtime")]
+    pub(crate) fn custom_code_memory(&self) -> Option<Arc<dyn CustomCodeMemory>> {
```
Would it be possible to return `Option<&dyn ...>` here to avoid forcing the clone on callers? Or perhaps `Option<&Arc<dyn ...>>` if callers should optionally have the ability to create an owned copy

alexcrichton commented on crates/wasmtime/src/runtime/vm/mmap_vec.rs:
```diff
@@ -35,6 +44,38 @@ pub enum MmapVec {
+mod base_ptr {
+    use core::ptr::NonNull;
+
+    /// A base pointer to an allocation, guaranteed not to be null,
+    /// and `Send`/`Sync`.
+    #[derive(Clone, Copy)]
+    pub struct BasePtr(NonNull<u8>);
```
Would `SendSyncPtr` suffice for this?

alexcrichton commented on crates/wasmtime/src/runtime/vm/mmap_vec.rs:
```diff
@@ -20,13 +22,20 @@ use std::fs::File;
-    Vec(Vec<u8>),
+    Alloc {
+        base: base_ptr::BasePtr,
+        len: usize,
```
I think `len` can be dropped here in favor of `layout.size()`?

cfallin replied:
Done!

alexcrichton commented on crates/wasmtime/src/compile/runtime.rs:
```diff
@@ -137,7 +137,7 @@ impl FinishedObject for MmapVecWrapper {
             fn reserve(&mut self, additional: usize) -> Result<(), ()> {
                 assert!(self.mmap.is_none(), "cannot reserve twice");
-                self.mmap = match MmapVec::with_capacity(additional) {
+                self.mmap = match MmapVec::with_capacity_and_alignment(additional, 1) {
```
I think this may not work for the use case of combining `CustomCodeMemory` and compilation at runtime, so the `Engine` and its custom alignment might need to be plumbed down here?

cfallin replied:
Done -- this was slightly gnarly to plumb through (required an extra `State` arg on the `FinishedObject` trait in `environ` and propagated through all the delegated builder stuff and cache handling and ...) but seems to work.

alexcrichton reviewed (approved):
Would it be possible to add a test exercising this trait as well? Perhaps x86_64-linux-specific to avoid rewriting a bunch of logic to make things executable or something like that.

cfallin commented:
Updated, thanks for feedback! Added a test that uses `rustix` on non-Windows targets to delegate to `mprotect` ("custom but actually the same as always" implementation).
